### PR TITLE
Fix #4255: Translate bitwise operations in query

### LINKED ISSUE #4255: Query: Translate bitwise operations
When I use flags into a model, bitwise operators could not be translate to sql with following code :
```
    public class MyModel
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public Week Days { get; set; }
    }

    [Flags]
    public enum Week
    {
        Monday = 1,
        Tuesday = 2,
        Wednesday = 4,
        Thursday = 8,
        Friday = 16,
        Saturday = 32,
        Sunday = 64
    }
```
Where clause exemple :
```
MyModels.Where(m => (m.Days & Week.Wednesday) > 0)
```
Debug message show :
```
Microsoft.Data.Entity.Query.Internal.SqlServerQueryCompilationContextFactory: Warning: The LINQ expression '((Convert([m.Days) & 4) > 0)' could not be translated and will be evaluated locally.
```
The issue cause bad performance because the request return all lines of the table before filtering.

For EF 6, same code works and the code has been translate to SQL.
```
SELECT 
    [Extent1].[Id] AS [Id], 
    [Extent1].[Days] AS [Days]
    FROM [dbo].[MyModels] AS [Extent1]
    WHERE  CAST( ( CAST( [Extent1].[Days] AS int)) & (4) AS int) > 0
```

### PR DESCRIPTION
**Pull request by fujiy**
Fix #4255

### CODE REVIEW

**Comment by smitpatel** [comment]
@anpete,@maumar, @divega  - this LGTM. If you want to review.
@fujiy - Can squash all commits in one and rebase on latest dev?

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/GearsOfWarQuerySqlServerTest.cs [hunk 1]
```diff
@@ -584,6 +584,88 @@ public override void Where_nullable_enum_with_nullable_parameter()
...
+
+        [ConditionalFact]
+        public override void Where_bitwise_and_nullable_enum_with_null_constant()
+        {
+            base.Where_bitwise_and_nullable_enum_with_null_constant();
+
+            Assert.Equal(
+                @"SELECT [w].[Id], [w].[AmmunitionType], [w].[IsAutomatic], [w].[Name], [w].[OwnerFullName], [w].[SynergyWithId]
+FROM [Weapon] AS [w]
+WHERE [w].[AmmunitionType] & NULL > 0",
```
Comment (divega) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/GearsOfWarQuerySqlServerTest.cs]: I am bit curious about how the behavior of bitwise operators will be with null in the database vs. in-memory, as a predicate but also in other places, e.g. the projection. I talked to @smitpatel and we will try it.

Comment (smitpatel) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/GearsOfWarQuerySqlServerTest.cs]: in-memory and database both working in same way for this. i.e. for int/long/enum values result is null when one of the operator is null. For values like string, it is compiler error (T-Sql also throws on that)

**Comment by smitpatel** [comment]
@fujiy - code changes look fine. Can you add few more tests?
Scenario we would like to cover
- Comparison of result of bitwise operator with enum value instead of `> 0`
- Bitwise operation on 1-2 other types like int or long (whatever types are available in current fixture is fine)
- Bitwise operation in projection instead of in where clause.

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/GearsOfWarQuerySqlServerTest.cs (hunk 1 above)
Comment (divega) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/GearsOfWarQuerySqlServerTest.cs]: Thanks. This is good information. So it seems it is good as it is.

**Comment by fujiy** [comment]
I can't run the unit tests after get the latest commits from dev. The tests output says:

Discovering tests in 'D:\Source\EntityFramework\test\Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests\project.json' ["C:\Program Files\dotnet\dotnet.exe" test "D:\Source\EntityFramework\test\Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests\project.json" --output "D:\Source\EntityFramework\artifacts\test\Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests\bin\Debug\netcoreapp1.0" --port 38116 --parentProcessId 29248 --no-build]
**'test-xunit' returned '-2147450749'.
'test-xunit' returned '-2147450749'.**
Discovering tests in 'D:\Source\EntityFramework\test\Microsoft.EntityFrameworkCore.SqlServer.Tests\project.json' ["C:\Program Files\dotnet\dotnet.exe" test "D:\Source\EntityFramework\test\Microsoft.EntityFrameworkCore.SqlServer.Tests\project.json" --output "D:\Source\EntityFramework\artifacts\test\Microsoft.EntityFrameworkCore.SqlServer.Tests\bin\Debug\netcoreapp1.0" --port 38129 --parentProcessId 29248 --no-build]
Discovering tests in 'D:\Source\EntityFramework\test\Microsoft.EntityFrameworkCore.Tests\project.json' ["C:\Program Files\dotnet\dotnet.exe" test "D:\Source\EntityFramework\test\Microsoft.EntityFrameworkCore.Tests\project.json" --output "D:\Source\EntityFramework\artifacts\test\Microsoft.EntityFrameworkCore.Tests\bin\Debug\netcoreapp1.0" --port 38130 --parentProcessId 29248 --no-build]
'test-xunit' returned '-2147450749'.
'test-xunit' returned '-2147450749'.
'test-xunit' returned '-2147450749'.
'test-xunit' returned '-2147450749'.

I tried to delete artifacts folder, rebuild several times, from VS and command line.

**Comment by fujiy** [comment]
Done :)

**Comment by smitpatel** [comment]
@maumar , @anpete - Any feedback on this? Else I will proceed with merge?

**Comment by smitpatel** [comment]
Merged in 7d21a8afb05a0f5a044de35ca0217743b9bd1c06
Thank you for contribution.

**Pull request closed without merging**